java-checks/src/main/java/org/sonar/java/checks/unused/UnusedPrivateFieldCheck.java

```diff
@@ -62,63 +68,120 @@
 
   private List<ClassTree> classes = Lists.newArrayList();
   private ListMultimap<Symbol, IdentifierTree> assignments = ArrayListMultimap.create();
+  private Set<String> unknownIdentifiers = new HashSet<>();
   private boolean hasNativeMethod = false;
 
   @Override
   public List<Kind> nodesToVisit() {
-    return ImmutableList.of(Tree.Kind.CLASS, Tree.Kind.COMPILATION_UNIT, Tree.Kind.METHOD, Tree.Kind.EXPRESSION_STATEMENT);
+    return ImmutableList.of(Tree.Kind.CLASS, Tree.Kind.METHOD, Tree.Kind.EXPRESSION_STATEMENT, Tree.Kind.IDENTIFIER);
   }
 
   @Override
-  public void leaveNode(Tree tree) {
-    if (hasSemantic()) {
-      if (tree.is(Tree.Kind.METHOD)) {
-        MethodTree method = (MethodTree) tree;
-        if (ModifiersUtils.hasModifier(method.modifiers(), Modifier.NATIVE)) {
-          hasNativeMethod = true;
-        }
-      } else if (tree.is(Tree.Kind.CLASS)) {
-        classes.add((ClassTree) tree);
-      } else if (tree.is(Tree.Kind.EXPRESSION_STATEMENT)) {
-        ExpressionTree expression = ((ExpressionStatementTree) tree).expression();
-        if (expression.is(ASSIGNMENT_KINDS)) {
-          addAssignment(((AssignmentExpressionTree) expression).variable());
-        }
-      } else {
-        leaveCompilationUnit();
-      }
-    }
-  }
-
-  private void leaveCompilationUnit() {
+  public void scanFile(JavaFileScannerContext context) {
+    super.scanFile(context);
     if (!hasNativeMethod) {
-      for (ClassTree classTree : classes) {
-        checkClassFields(classTree);
-      }
+      classes.forEach(this::checkClassFields);
     }
     classes.clear();
     assignments.clear();
+    unknownIdentifiers.clear();
     hasNativeMethod = false;
   }
 
-  private void checkClassFields(ClassTree classTree) {
-    for (Tree member : classTree.members()) {
-      if (member.is(Tree.Kind.VARIABLE)) {
-        checkIfUnused((VariableTree) member);
-      }
+  @Override
+  public void visitNode(Tree tree) {
+    if (!hasSemantic()) {
+      return;
+    }
+    switch (tree.kind()) {
+      case METHOD:
+        checkIfNativeMethod((MethodTree) tree);
+        break;
+      case CLASS:
+        classes.add((ClassTree) tree);
+        break;
+      case EXPRESSION_STATEMENT:
+        collectAssignment(((ExpressionStatementTree) tree).expression());
+        break;
+      case IDENTIFIER:
+        collectUnknownIdentifier((IdentifierTree) tree);
+        break;
+      default:
+        throw new IllegalStateException("Unexpected subscribed tree.");
+    }
+  }
+
+  private void collectUnknownIdentifier(IdentifierTree identifier) {
+    if (!isVariableIdentifier(identifier) && !isMethodIdentifier(identifier) && identifier.symbol().isUnknown()) {
```

Comment:
OK for first part.
For the second part: I'm not quite sure if it's a bug or a feature... From the `VariableTree`, you can access the variable symbol (using `symbol()`), but if you try to get the symbol from the `IdentifierTree` corresponding to the `simpleName()` of the `VariableTree`, you get _unknown_.